src/encoded/types/workflow.py

```diff
@@ -135,6 +140,367 @@
+            sorted_wfr_tuples = sorted(wfr_tuples_for_wf, key=get_date_created_from_tuple)
+            filtered_tuples.append(sorted_wfr_tuples[0])
+            filtered_out_tuples = filtered_out_tuples + sorted_wfr_tuples[1:]
+
+        return (filtered_tuples, filtered_out_tuples)
+
+
+    def generate_sources_for_input(in_files, workflow_argument_name, depth = 0):
+
+        sources = [] # Our output
+        step_uuids = set()
+
+        def try_match_input_with_workflow_run_output_to_generate_source(workflow_run_model_obj, workflow_run_uuid, in_file, in_file_uuid):
+            sources_for_in_file = [] # We only are looking for 1 source, but might re-use for trace_future later
+            for out_file in workflow_run_model_obj.get('output_files', []):
+                out_file_atid = out_file.get('value', {})
+                if out_file_atid == in_file.get('@id', 'b'):
+                    step_name = workflow_run_model_obj.get('display_title')
+                    step_uuid = workflow_run_uuid
+                    if step_uuid:
+                        step_uuids.add(step_uuid)
+                    sources_for_in_file.append({
+                        "name" : out_file.get('workflow_argument_name'),
+                        "step" : step_name,
+                        "type" : "Output file",
+                        "for_file" : in_file_uuid,
+                        "workflow" : workflow_run_model_obj.get('workflow')
+                    })
+            return sources_for_in_file
+
+        # Gather all workflow_runs out of which our input files (1 run per file) come from
+        all_workflow_runs = []
+
+        for in_file in in_files:
+            in_file_uuid = in_file.get('uuid')
+            if uuidCacheTracedHistory.get(in_file_uuid):
+                sources = sources + uuidCacheTracedHistory[in_file_uuid]
+                continue
+
+            input_file_model = get_model_by_uuid(in_file_uuid)
+
+            if not input_file_model or not hasattr(input_file_model, 'source') or input_file_model.source.get('object') is None:
+                continue
+
+            input_file_model_obj = input_file_model.source.get('object', {})
+
+            # Update in_file with metadata we want from the file.
+            in_file['@type'] = input_file_model_obj.get('@type')
+            in_file['file_type'] = input_file_model_obj.get('file_type')
+            in_file['filename'] = input_file_model_obj.get('filename')
+            # Get @ids from ES source.
+            output_of_workflow_runs = input_file_model_obj.get('workflow_run_outputs', [])
+            if len(output_of_workflow_runs) == 0:
+                continue
+            # There should only ever be one 'workflow_run_outputs' at most, or versions of same one (grab most recent).
+            last_workflow_run_output_of = output_of_workflow_runs[len(output_of_workflow_runs) - 1]
+            workflow_run_uuid = get_unique_key_from_at_id(last_workflow_run_output_of)
```

Comment:
seems like these four lines could all be part of get_model_by_uuid .. maybe just call it get_model_by_atid?

Reply:
I need to rename that func lol. Reasoning I kept it explicit is that sometimes can grab a UUID from atid, sometimes can only grab accesssion. When use accession, need to use `get_by_unique_key(accession, 'accession')` rather `than get_by_uuid`, and get_by_unique_key returns a different type of Item (Snovault Item I believe) than get_by_uuid, which returns a model. If don't make clear distinction (uuid vs accession), then might become more confusing to determine if are dealing with a SnowyItem (via accession) or ORM model (via UUID). Eventually I'm thinking could just make it request.embed() (the `get_by_atid`) once/if performance improves and delete a lot of lines of code?

Reply:
Oops nevermind.... I thought GH comments were about the lines above the comment not below. Yeah definitely, will do.

Reply:
Actually... I do seem to be getting model from get_by_unique_key, well then, this simplifies things.... :-D